### PR TITLE
Replace getProject() references with injected services in task implementations where possible

### PR DESCRIPTION
**Pull request by breskeby**
Usages of `getProject()` within a task should be avoided as it is not compatible
with gradle configuration cache. Related to #57918

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java [hunk 1]
```diff
@@ -177,25 +207,14 @@ public void ignoreJarHellWithJDK(String... classes) {
         // err on the side of scanning these to make sure we don't miss anything
         Spec<Dependency> reallyThirdParty = dep -> dep.getGroup() != null && dep.getGroup().startsWith("org.elasticsearch") == false;
         Set<File> jars = getRuntimeConfiguration().getResolvedConfiguration().getFiles(reallyThirdParty);
-        Set<File> compileOnlyConfiguration = getProject().getConfigurations()
```
Comment (pugnascotia) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java:180]: Was this code not actually required?

**Review by pugnascotia — APPROVED** [review]

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java [hunk 2]
```diff
@@ -35,22 +40,29 @@
 public class NoticeTask extends DefaultTask {
 
     @InputFile
-    private File inputFile = getProject().getRootProject().file("NOTICE.txt");
+    private File inputFile;
+
     @OutputFile
-    private File outputFile = new File(getProject().getBuildDir(), "notices/" + getName() + "/NOTICE.txt");
+    private File outputFile;
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java:46]: Can we put a linebreak in here so this is more readable?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java (hunk 1 above)
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java:180]: runtimeConfiguration should not contain any `compileOnly` configurations

File: build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/precommit/PomValidationTask.java [hunk 3]
```diff
@@ -21,12 +22,19 @@
...
 public class PomValidationTask extends PrecommitTask {
 
-    private final RegularFileProperty pomFile = getProject().getObjects().fileProperty();
+    private final RegularFileProperty pomFile;
 
     private boolean foundError;
 
+    @Inject
+    public PomValidationTask(ObjectFactory objects) {
+        pomFile= objects.fileProperty();
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/precommit/PomValidationTask.java:35]: we do not apply checkstyle on build-convention, due to chicken egg. fixed the formatting

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java [hunk 4]
```diff
@@ -35,22 +40,29 @@
...
     private List<File> licensesDirs = new ArrayList<File>();
+    private final FileOperations fileOperations;
 
-    public NoticeTask() {
+    @Inject
+    public NoticeTask(BuildLayout buildLayout, ProjectLayout projectLayout, FileOperations fileOperations) {
         setDescription("Create a notice file from dependencies");
         // Default licenses directory is ${projectDir}/licenses (if it exists)
-        File licensesDir = new File(getProject().getProjectDir(), "licenses");
+        File licensesDir = projectLayout.getProjectDirectory().dir("licenses").getAsFile();
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java:59]: done

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java (hunk 2 above)
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java:46]: done

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java [hunk 5]
```diff
@@ -35,22 +40,29 @@
...
     @OutputFile
-    private File outputFile = new File(getProject().getBuildDir(), "notices/" + getName() + "/NOTICE.txt");
+    private File outputFile;
     private FileTree sources;
+
     /**
      * Directories to include notices from
      */
     private List<File> licensesDirs = new ArrayList<File>();
+    private final FileOperations fileOperations;
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/NoticeTask.java:53]: done

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DependenciesInfoTask.java [hunk 6]
```diff
@@ -101,7 +106,12 @@ public void setOutputFile(File outputFile) {
     @InputFiles
     private Configuration compileOnlyConfiguration;
 
-    public DependenciesInfoTask() {
+    @Inject
+    public DependenciesInfoTask(ProjectLayout pLayout) {
+        this.projectLayout = pLayout;
+        File dir = pLayout.getProjectDirectory().dir("licenses").getAsFile();
+        this.licensesDir = dir.exists() ? dir : null;
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DependenciesInfoTask.java:113]: done

**Review by mark-vieira — APPROVED** [review]

**Comment by mark-vieira** [comment]
@breskeby I don't think this got backported to the `7.17` branch. Can you do that and ensure the fix in https://github.com/elastic/elasticsearch/pull/84310 is incorporated into the backport?

**Comment by breskeby** [comment]
@mark-vieira done

**Pull request merged**